learning_observer/learning_observer/learning_tools/canvas.py

```diff
@@ -0,0 +1,162 @@
+import json
+import requests
+import configparser
+
+class Canvas:
+    def __init__(self, config_path='./config.ini'):
+        self.config = configparser.ConfigParser()
+        self.config.read(config_path)
+        self.defaultServer = self.config['CANVAS_CONFIG']['DEFAULT_SERVER']
+        self.access_token = self.config['CANVAS_CONFIG']['ACCESS_TOKEN']
+        self.refresh_token = self.config['CANVAS_CONFIG']['REFRESH_TOKEN']
+        self.client_id = self.config['CANVAS_CONFIG']['CLIENT_ID']
+        self.client_secret = self.config['CANVAS_CONFIG']['CLIENT_SECRET']
+        self.default_version = 'v1'
+        self.defaultPerPage = 10000
+        self.base_url = f'https://{self.defaultServer}/api/{self.default_version}'
+
+    def update_access_tokens(self, access_token):
+        self.config['CANVAS_CONFIG']['ACCESS_TOKEN'] = access_token
+        self.access_token = access_token
+        with open('./config.ini', 'w') as configfile:
+            self.config.write(configfile)
+
+    def api_call(self, method, endpoint, params=None, data=None, absolute_url=False, retry=True):        
+        if absolute_url:
+            url = endpoint
+        else:
+            url = self.base_url + endpoint
+            # Append params key/values if added
+            if params:
+                url += '?' + '&'.join(f"{k}={v}" for k, v in params.items())
```

Comment:
This code looks a little familiar to me. I believe there may be a utility function floating around somewhere for formatting the parameters of a url. Spend a little bit of time looking for that function both on the internet and within the codebase, but don't fret if you can't find it.
It's best to remain consistent in how we do things.